src/backends/perf_pt/mod.rs

```diff
@@ -58,6 +57,55 @@ extern "C" {
     fn perf_pt_free_tracer(tr_ctx: *const c_void) -> c_int;
 }
 
+/// A Intel PT trace, obtained via Linux Perf.
+#[allow(dead_code)]
+pub struct PerfPTTrace {
+    buf: *const u8,
+    len: u64,
+}
+
+impl PerfPTTrace {
+    /// Makes a new trace from a raw pointer and a size.
+    ///
+    /// The `buf` argument is assumed to have been allocated on the heap using malloc(3). `len`
+    /// must be less than or equal to the allocated size.
+    ///
+    /// The allocation is automatically freed by Rust when the struct falls out of scope.
+    fn from_buf(buf: *const u8, len: u64) -> Self {
+        Self {buf: buf, len: len}
+    }
+
+}
+
+impl Trace for PerfPTTrace {
+    /// Write the raw trace packets into the specified file.
+    ///
+    /// This can be useful for developers who want to use (e.g.) the pt utility to inspect the raw
+    /// packet stream.
+    #[cfg(debug_assertions)]
+    fn to_file(&self, filename: &str) {
+        use std::slice;
+        use std::fs::File;
+        use std::io::prelude::*;
+
+        let mut f = File::create(filename).unwrap();
+        let slice = unsafe { slice::from_raw_parts(self.buf, self.len as usize) };
+        f.write(slice).unwrap();
+    }
+}
+
+
+/// Once a PerfPTTrace is brought into existence, we say the instance owns the C-level allocation.
+/// When the it falls out of scope, free up the memory.
+impl Drop for PerfPTTrace {
+    fn drop(&mut self) {
+        if self.buf != ptr::null() {
```

Comment:
How could `self.buf` be `null`?

Reply:
If `start_tracing` or `stop_tracing` failed for example.

https://github.com/softdevteam/hwtracer/blob/9902d858002e5bd3a4e599f97ea8c7f3a7285122/src/backends/perf_pt/perf_pt.c#L488

https://github.com/softdevteam/hwtracer/blob/9902d858002e5bd3a4e599f97ea8c7f3a7285122/src/backends/perf_pt/perf_pt.c#L542

Reply:
Maybe I'm an idiot, but at least the first of those seems to be writing to a value that isn't `self.buf`?

Reply:
It's the same buffer.

The buffer first lives in ` tr_ctx->trace_buf` until the user stops tracing, then "ownership" is transferred to the Rust `PerfPTTrace` struct here:

https://github.com/softdevteam/hwtracer/blob/9902d858002e5bd3a4e599f97ea8c7f3a7285122/src/backends/perf_pt/perf_pt.c#L534

Where `buf` is a pointer to the `self.buf` on the Rust side.

[This relates to your other question about allocations]

Reply:
OK, I admit that the split between Rust and C code makes this a bit hard to understand in isolation.